_posts/2025-02-21-katib-rag-optimization.md

```diff
@@ -0,0 +1,283 @@
+retriever_model = SentenceTransformer('paraphrase-MiniLM-L6-v2')
+
+documents = fetch_documents()
+doc_embeddings = retriever_model.encode(documents)
+
+# Build FAISS index
+index = faiss.IndexFlatL2(doc_embeddings.shape[1])
+index.add(np.array(doc_embeddings))
+
+# Generator: Pre-trained GPT-2
+generator = pipeline("text-generation", model="gpt2", tokenizer="gpt2")
+
+
+def rag_pipeline_execute(query, top_k=1, temperature=1.0):
+   query_embedding = retriever_model.encode([query])
+   distances, indices = index.search(query_embedding, top_k)
+   retrieved_docs = [documents[i] for i in indices[0]]
+
+   context = " ".join(retrieved_docs)
+   generated = generator(context, max_length=50, temperature=temperature, num_return_sequences=1)
+   return generated[0]['generated_text']
+
+
+def fetch_documents():
+   return [
+      # Retun the list of documents...
+   ]
+```
+
+#### Evaluating the Pipeline - BLEU Score Calculation
+
+The `evaluate` function measures the quality of generated text using the BLEU
+score. This function runs the RAG pipeline for a given query, compares the
+output against a reference answer, and computes the BLEU score with smoothing
+techniques.
+
+```python
+import argparse
+from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
+from rag import rag_pipeline_execute
+
+# Simulated RAG pipeline (simplified for example)
+def rag_pipeline(query, top_k, temperature):
+   return rag_pipeline_execute(query, top_k=top_k, temperature=temperature)
+
+
+# Evaluate BLEU score
+def evaluate(query, ground_truth, top_k, temperature):
+   # Get the RAG pipeline response
+   response = rag_pipeline(query, top_k, temperature)
+
+   # Tokenize the response and ground truth
+   reference = [ground_truth.split()]  # Reference should be a list of tokens
+   candidate = response.split()  # Candidate is the generated response tokens
+
+   # Apply smoothing to the BLEU score
+   smoothie = SmoothingFunction().method1  # Use method1 for smoothing
+   bleu_score = sentence_bleu(reference, candidate, smoothing_function=smoothie)
+
+   return bleu_score
+```
+
+## STEP 3: Run a Katib Experiment
+
+To optimize the RAG pipeline’s hyperparameters, we need to submit an experiment to Katib, which will automatically 
+search for the best configuration. We can do this using the [Katib SDK][Katib_SDK].
+
+The first step would be to define an `objective` function which is what we will optimize. It runs the RAG pipeline with 
+different hyperparameters and logs the BLEU score. 
+
+```python
+# Objective function for Katib
+def objective(parameters):
+    # TODO: Provide queries and ground truth directly here or load them dynamically from a file/external volume.
+    query = ""  # Example: "Tell me about the Eiffel Tower."
+    ground_truth = ""  # Example: "The Eiffel Tower is a famous landmark in Paris."
+
+    # Run the RAG pipeline with hyperparameters from Katib
+    bleu_score = evaluate(query, ground_truth, top_k, temperature)
```

Comment:
We should say that all imports and functions (e.g. `evaluate()`) should be defined under the `objective` function.
Similar to KFP Lightweight Python Component, the function must be self-contained: https://www.kubeflow.org/docs/components/pipelines/user-guides/components/lightweight-python-components/#python-function-requirements.

Users can also do something like this:
```python
def objective(parameters):
  from evaluate import eval
```

But in that case, evaluate module should be accessible in Katib Trials.

Reply:
I've restructured the content to talk about the objective function a clearly now